staging/src/k8s.io/apiserver/pkg/util/flowcontrol/conc_alloc.go

```diff
@@ -0,0 +1,252 @@
+			return nil, 0, fmt.Errorf("upper bound %d is %v but should not be less than the lower bound %v", idx, item.upperBound, item.lowerBound)
+		}
+		if target < MinTarget {
+			// tweak this to a non-zero value so avoid dividing by zero
+			target = MinTarget
+		}
+		lowSum += item.lowerBound
+		highSum += item.upperBound
+		targetSum += target
+		relativeItem := relativeAllocItem{
+			target:             target,
+			relativeLowerBound: item.lowerBound / target,
+			relativeUpperBound: item.upperBound / target,
+		}
+		ubRange.note(relativeItem.relativeUpperBound)
+		lbRange.note(relativeItem.relativeLowerBound)
+		relativeItems[idx] = relativeItem
+	}
+	if lbRange.max > 1 {
+		panic(lbRange)
+	}
+	if lowSum-requiredSumF > epsilon {
+		return nil, 0, fmt.Errorf("lower bounds sum to %v, which is higher than the required sum of %v", lowSum, requiredSum)
+	}
+	if requiredSumF-highSum > epsilon {
+		return nil, 0, fmt.Errorf("upper bounds sum to %v, which is lower than the required sum of %v", highSum, requiredSum)
+	}
+	ans := make([]float64, len(classes))
+	if requiredSum == 0 {
+		return ans, 0, nil
+	}
+	if lowSum-requiredSumF > -epsilon { // no wiggle room, constrained from below
+		for idx, item := range classes {
+			ans[idx] = item.lowerBound
+		}
+		return ans, lbRange.min, nil
+	}
+	if requiredSumF-highSum > -epsilon { // no wiggle room, constrained from above
+		for idx, item := range classes {
+			ans[idx] = item.upperBound
+		}
+		return ans, ubRange.max, nil
+	}
+	// Now we know the solution is a unique fairProp in [lbRange.min, ubRange.max].
+	// See if the solution does not run into any bounds.
+	fairProp := requiredSumF / targetSum
+	if lbRange.max <= fairProp && fairProp <= ubRange.min { // no bounds matter
+		for idx := range classes {
+			ans[idx] = relativeItems[idx].target * fairProp
+		}
+		return ans, fairProp, nil
+	}
+	// Sadly, some bounds matter.
+	// We find the solution by sorting the bounds and considering progressively
+	// higher values of fairProp, starting from lbRange.min.
+	rap := (&relativeAllocProblem{items: relativeItems}).initIndices()
+	sumSoFar := lowSum
+	fairProp = lbRange.min
+	var weight, deltaWeight float64
+	var weighing, deltaWeighing int
+	lastIdx := -1
+	// `lastIdx` is the last `rap` index that has been considered.
+	// `sumSoFar` is what the allocs would sum to if the current
+	// value of `fairProp` solves the problem.
+	// If the current value of fairProp were the answer then
+	// `sumSoFar == requiredSum`.
+	// Otherwise the next increment in fairProp involves changing the allocations
+	// of `weighing` classes whose targets sum to `weight`.
+	// Raising fairProp above `lastidx` adds `deltaWeight` to `weight`
+	// and `deltaWeighing` to `weighing`.
+	for sumSoFar < requiredSumF {
+		// Find the next bound that constrains fairProp.
+		var nextBound float64
+		for {
+			lastIdx++
+			weight += deltaWeight
+			weighing += deltaWeighing
+			if lastIdx >= rap.Len() {
+				panic(*rap)
```

Comment:
Can we avoid panic?

Let's do something like:
```
return fmt.Errorf("programming error: ...")
```

[I know that by design this should never happen, but in case of a bug, I don't want my apiserver to crash...]

Reply:
done